include/barrier.h

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2022 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public interface for barriers.
+ */
+#ifndef ZEPHYR_INCLUDE_BARRIER_H_
+#define ZEPHYR_INCLUDE_BARRIER_H_
+
+#ifndef _ASMLANGUAGE
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup barrier_apis Barriers API
+ * @ingroup kernel_apis
+ * @{
+ */
+
+#if !defined(nop) || defined(__DOXYGEN__)
+/**
+ * @brief nop
+ *
+ * NOP instruction.
+ */
+#define nop()		__asm__ volatile ("nop")
+#endif /* !nop */
+
+#if !defined(isb) || defined(__DOXYGEN__)
+/**
+ * @brief isb
+ *
+ * This is used to guarantee that any subsequent instructions are fetched, so
+ * that privilege and access are checked with the current MMU configuration.
+ */
+#define isb()
+#endif /* !isb */
+
+#if !defined(mb) || defined(__DOXYGEN__)
+/**
+ * @brief mb - Memory Barrier
+ *
+ * A full system memory barrier. All memory operations before the mb() in the
+ * instruction stream will be committed before any operations after the mb()
+ * are committed.
+ */
+#define mb()		compiler_barrier()
+#endif /* !mb */
+
+#if !defined(rmb) || defined(__DOXYGEN__)
+/**
+ * @brief rmb - Read Memory Barrier
+ *
+ * Like mb(), but only guarantees ordering between read accesses. That is, all
+ * read operations before an rmb() will be committed before any read operations
+ * after the rmb().
+ */
+#define rmb()		mb()
+#endif /* !rmb */
+
+#if !defined(wmb) || defined(__DOXYGEN__)
+/**
+ * @brief wmb - Write Memory Barrier
+ *
+ * Like mb(), but only guarantees ordering between write accesses. That is, all
+ * write operations before a wmb() will be committed before any write
+ * operations after the wmb().
+ */
+#define wmb()		mb()
+#endif /* !wmb */
+
+#if defined(CONFIG_SMP) || defined(__DOXYGEN__)
+
+#if !defined(smp_mb) || defined(__DOXYGEN__)
+/**
+ * @brief smp_mb - SMP Memory Barrier
+ *
+ * Similar to mb(), but only guarantees ordering between cores/processors
+ * within an SMP system. All memory accesses before the smp_mb() will be
+ * visible to all cores within the SMP system before any accesses after the
+ * smp_mb().
+ */
```

Comment:
Maybe "local" would be clearer than "smp" here?  Having a prefix like this would seem to imply that this is what you "should" use on a multiprocessor system.  But in most cases you shouldn't, this is an optimized version you **can** use if you really know what you're doing.

Frankly this looks like a footgun to me.  Does it really help on actual hardware?  I can imagine a whole generation of developers coming through and deciding they want the "SMP version" and then dealing with subtle breakage in their DMA drivers for decades to come.  Maybe just drop it unless there's a really clear use case?

Reply:
Uhm, the `*_smp` version is indeed a "lighter" version. The problem is that currently in Zephyr we have a wide usage of both `__DSB` and `__DMB` (why?) , so the `*_smp` version could be used where `__DMB` is currently used.